### PR TITLE
Netty changes for pipeline change

### PR DESCRIPTION
**Pull request by ashwinr89**

### CODE REVIEW

**Comment by ashwinr89** [comment]
@krdev Can you review?

File: src/main/java/com/lafaspot/pop/client/PopClient.java [hunk 1]
```diff
@@ -40,8 +40,9 @@
     /** Event loop group that will serve all channels for IMAP client. */
     private final EventLoopGroup group;
 
-    /** The log manger. */
-    private final LogManager logManager;
+    /** The pop session handling the outgoing session. */
+    private PopSession session;
```
Comment (krdev) [inline src/main/java/com/lafaspot/pop/client/PopClient.java]: remove

File: src/main/java/com/lafaspot/pop/session/PopSession.java [hunk 2]
```diff
@@ -75,7 +92,18 @@ public PopSession(@Nonnull final SslContext sslContext, @Nonnull final Bootstrap
     }
 
     /**
+     * Returns the channel that the pop client is using to connect to the remote server.
+     *
+     * @return the destination channel.
+     */
+    public Channel getDestinationChannel() {
```
Comment (krdev) [inline src/main/java/com/lafaspot/pop/session/PopSession.java]: rename to getChannel

File: src/main/java/com/lafaspot/pop/client/PopClient.java [hunk 3]
```diff
@@ -58,7 +59,6 @@
      */
     public PopClient(final int threads, @Nonnull final LogManager logManager) throws PopException {
```
Comment (krdev) [inline src/main/java/com/lafaspot/pop/client/PopClient.java:58]: Use Properties class pass in auto-read as a config, pass it in the PopClient c'tor (2nd c'tor)

File: src/main/java/com/lafaspot/pop/client/PopClient.java [hunk 4]
```diff
@@ -68,7 +66,36 @@ public PopClient(final int threads, @Nonnull final LogManager logManager) throws
...
+        this.group = new NioEventLoopGroup(threads);
+        try {
+            this.sslContext = SslContextBuilder.forClient().build();
+            this.bootstrap.group(this.group);
+            this.bootstrap.channel(NioSocketChannel.class);
+            this.bootstrap.option(ChannelOption.SO_KEEPALIVE, true);
+            final Iterator<Object> it = properties.keySet().iterator();
+            while (it.hasNext()) {
+                final ChannelOption obj = (ChannelOption) it.next();
+                this.bootstrap.option(obj, properties.get(obj));
```
Comment (krdev) [inline src/main/java/com/lafaspot/pop/client/PopClient.java]: Have well defined property values that you look for and pass down.

File: src/main/java/com/lafaspot/pop/session/PopSession.java [hunk 5]
```diff
@@ -116,10 +138,10 @@ protected void initChannel(final SocketChannel ch) throws Exception {
             throw new PopException(Type.CONNECT_FAILURE, e);
         }
 
-        stateRef.compareAndSet(State.NULL, State.COMMAND_SENT);
```
Comment (krdev) [inline src/main/java/com/lafaspot/pop/session/PopSession.java:119]: do the check + set and fail also, move the check before actual network operation, same for all other cases.

File: src/main/java/com/lafaspot/pop/session/PopSession.java [hunk 6]
```diff
@@ -162,10 +184,10 @@ public void operationComplete(final Future<? super Void> future) throws Exceptio
      * @throws PopException on failure
      */
     public PopFuture<PopCommandResponse> disconnect() throws PopException {
-    	final State state = stateRef.get();
-    	if (state == State.NULL) {
+        final State state = stateRef.get();
+        if (state == State.NULL) {
```
Comment (krdev) [inline src/main/java/com/lafaspot/pop/session/PopSession.java]: check for connected state before disconnect

File: src/test/java/com/lafaspot/pop/client/PopClientProdIT.java [hunk 7]
```diff
@@ -47,6 +55,74 @@ public void init() throws PopException {
     }
 
     @Test
+    public void testChannelPipeline() throws PopException, InterruptedException, ExecutionException {
+        System.out.println("testChannelPipeline Begin");
```
Comment (krdev) [inline src/test/java/com/lafaspot/pop/client/PopClientProdIT.java]: only use our logging framework, remove all system.out.println

File: src/test/java/com/lafaspot/pop/client/PopClientProdIT.java [hunk 8]
```diff
@@ -47,6 +55,74 @@ public void init() throws PopException {
...
+        final Iterator<Entry<String, ChannelHandler>> it = pipe.iterator();
+        int index = 0;
+        while (it.hasNext()) {
+            final Entry<String, ChannelHandler> entry = it.next();
+            Assert.assertNotNull(entry);
+            final String name = entry.getKey();
+            Assert.assertEquals(name, handlerList[index++]);
+            Assert.assertNotNull(entry.getValue());
+        }
+        System.out.println("testChannelPipeline DONE");
```
Comment (krdev) [inline src/test/java/com/lafaspot/pop/client/PopClientProdIT.java]: complete the test properly by doing a disconnect, validating that disconnect worked properly

File: src/test/java/com/lafaspot/pop/client/PopClientProdIT.java [hunk 9]
```diff
@@ -47,6 +55,74 @@ public void init() throws PopException {
...
+        final String[] newHandlerList = { PopSession.SSL_HANDLER, "NEW_HANDLER" };
+        final Iterator<Entry<String, ChannelHandler>> it2 = pipe.iterator();
+        int index2 = 0;
+        while (it2.hasNext()) {
+            final Entry<String, ChannelHandler> entry = it2.next();
+            Assert.assertNotNull(entry);
+            final String name = entry.getKey();
+            Assert.assertEquals(name, newHandlerList[index2++]);
+            Assert.assertNotNull(entry.getValue());
+        }
```
Comment (krdev) [inline src/test/java/com/lafaspot/pop/client/PopClientProdIT.java:114]: the remove old + add new will never be atomic, test and  handle the case the following case:
- remove old
- channel disconnected by now
- you add new handlers

File: src/main/java/com/lafaspot/pop/session/PopSession.java (hunk 6 above)
Comment (ashwinr89) [reply src/main/java/com/lafaspot/pop/session/PopSession.java]: Check if state is not CONNECTED and throw.

File: src/main/java/com/lafaspot/pop/session/PopSession.java (hunk 5 above)
Comment (ashwinr89) [reply src/main/java/com/lafaspot/pop/session/PopSession.java:119]: Moved the check+set and fail fast to the beginning.

File: src/main/java/com/lafaspot/pop/client/PopClient.java (hunk 4 above)
Comment (ashwinr89) [reply src/main/java/com/lafaspot/pop/client/PopClient.java]: comparing the string from the properties keyset with a known set of keys.

File: src/main/java/com/lafaspot/pop/client/PopClient.java (hunk 3 above)
Comment (ashwinr89) [reply src/main/java/com/lafaspot/pop/client/PopClient.java:58]: Passing the config in the constructor as Properties.

File: src/test/java/com/lafaspot/pop/client/PopClientProdIT.java (hunk 8 above)
Comment (ashwinr89) [reply src/test/java/com/lafaspot/pop/client/PopClientProdIT.java]: Added the disconnect verification.

File: src/test/java/com/lafaspot/pop/client/PopClientProdIT.java (hunk 9 above)
Comment (ashwinr89) [reply src/test/java/com/lafaspot/pop/client/PopClientProdIT.java:114]: Adding new hanlders doesn't affect. We should be careful inside the handler to handle inactive channels.

Comment (krdev) [reply src/test/java/com/lafaspot/pop/client/PopClientProdIT.java:114]: where are the UTs I asked for?

Comment (ashwinr89) [reply src/test/java/com/lafaspot/pop/client/PopClientProdIT.java:114]: As previously stated, the the UTs for the use-case that you asked for, doesn't have any effect. So, it ll be the same as the previous UT.

**Pull request merged**